File: packages/protocol/src/evaluator.ts
import { encodeAbiParameters, keccak256, type Hex } from "viem";
import { ACTION_FTESTXRP_TRANSFER, ZERO_BYTES32 } from "./constants.js";
import { actionRequestHash, normalizePolicy, policyCommitment } from "./codec.js";
import type { ActionRequestV1, Decision, EvaluationResultV1, PolicyV1, PublicReasonClass, SpendStateV1 } from "./types.js";

function deny(request: ActionRequestV1, reason: PublicReasonClass, now: bigint, machineId: Hex = ZERO_BYTES32, keyFingerprint: Hex = ZERO_BYTES32): EvaluationResultV1 {
  return { request, decision: "DENY", publicReasonClass: reason, reservedAmount: 0n, resultingCheckpoint: request.spendCheckpoint,
    resultNonce: request.requestId, attempt: request.attempt, issuedAt: now, expiry: request.expiry, machineId, keyFingerprint };
}

function containsAddress(values: Hex[], value: Hex): boolean {
  return values.some((candidate) => candidate.toLowerCase() === value.toLowerCase());
}

function containsBytes32(values: Hex[], value: Hex): boolean {
  return values.some((candidate) => candidate.toLowerCase() === value.toLowerCase());
}

function checkedReferenceValue(amount: bigint, price: bigint, priceDecimals: number): bigint | null {
  if (price <= 0n || priceDecimals < 0 || priceDecimals > 36) return null;
  const scale = 10n ** BigInt(priceDecimals);
  if (amount !== 0n && price > ((1n << 256n) - 1n) / amount) return null;
  return (amount * price + scale - 1n) / scale;
}

function nextCheckpoint(request: ActionRequestV1, amount: bigint, occurrence: number, now: bigint): Hex {
  return keccak256(encodeAbiParameters(
    [{ type: "bytes32" }, { type: "bytes32" }, { type: "uint256" }, { type: "uint32" }, { type: "uint64" }],
    [request.spendCheckpoint, actionRequestHash(request), amount, occurrence, now],
  ));
}

export function evaluatePolicy(policyInput: PolicyV1, request: ActionRequestV1, state: SpendStateV1): EvaluationResultV1 {
  const policy = normalizePolicy(policyInput);
  const now = state.now;
  if (state.availableBalance < 0n || state.dailySpend < 0n || state.rollingSpend < 0n || state.occurrenceCount < 0) {
    return deny(request, "MALFORMED", now);
  }
  const domainMatches = request.chainId === policy.chainId && request.registry.toLowerCase() === policy.registry.toLowerCase()
    && request.vault.toLowerCase() === policy.vault.toLowerCase() && request.router.toLowerCase() === policy.router.toLowerCase()
    && request.policyId.toLowerCase() === policy.policyId.toLowerCase() && request.policyVersion === policy.policyVersion
    && request.policyCommitment.toLowerCase() === policyCommitment(policy).toLowerCase();
  if (!domainMatches) return deny(request, "WRONG_DOMAIN", now);
  if (state.spendCheckpoint.toLowerCase() !== request.spendCheckpoint.toLowerCase()
    || state.balanceCheckpoint.toLowerCase() !== request.balanceCheckpoint.toLowerCase()) {
    return deny(request, "STALE_INPUT", now);
  }
  if (request.asset.toLowerCase() !== policy.asset.toLowerCase() || request.actionType.toLowerCase() !== ACTION_FTESTXRP_TRANSFER.toLowerCase()
    || request.amount <= 0n || request.createdAt > now || request.expiry < now || request.graceDeadline < request.createdAt || request.expiry < request.graceDeadline) {
    return deny(request, request.expiry < now ? "EXPIRED" : "MALFORMED", now);
  }
  if (policy.startAt > now || (policy.endAt !== 0n && now > policy.endAt)) return deny(request, "POLICY_DENIED", now);
  if (containsAddress(policy.denyTargets, request.target)) return deny(request, "TARGET_DENIED", now);
  if (policy.allowTargets.length > 0 && !containsAddress(policy.allowTargets, request.target)) return deny(request, "TARGET_DENIED", now);
  if (policy.allowRequesters.length > 0 && !containsAddress(policy.allowRequesters, request.requester)) return deny(request, "REQUESTER_DENIED", now);
  if (policy.allowActionTypes.length > 0 && !containsBytes32(policy.allowActionTypes, request.actionType)) return deny(request, "ACTION_DENIED", now);
  if (policy.maxOccurrences !== 0 && state.occurrenceCount >= policy.maxOccurrences) return deny(request, "OCCURRENCE_EXCEEDED", now);
  if (policy.cooldownSeconds !== 0n && state.lastExecutionAt !== 0n && now < state.lastExecutionAt + policy.cooldownSeconds) return deny(request, "COOLDOWN", now);
  if (state.availableBalance < request.amount) return deny(request, "INSUFFICIENT_BALANCE", now);
  let referenceValue = request.amount;
  if (policy.requireFtso) {
    const feed = state.ftso;
    if (!feed || feed.feedId.toLowerCase() !== policy.ftsoFeedId.toLowerCase() || feed.timestamp > now || now - feed.timestamp > policy.maxPriceAgeSeconds
      || feed.checkpoint === ZERO_BYTES32 || request.inputCommitment.toLowerCase() !== feed.checkpoint.toLowerCase()) return deny(request, "FTSO_INVALID", now);
    const value = checkedReferenceValue(request.amount, feed.value, feed.decimals);
    if (value === null) return deny(request, "FTSO_INVALID", now);
    referenceValue = value;
  }
  if ((policy.maxPerAction !== 0n && referenceValue > policy.maxPerAction)
    || (policy.dailyCap !== 0n && state.dailySpend + referenceValue > policy.dailyCap)
    || (policy.rollingCap !== 0n && state.rollingSpend + referenceValue > policy.rollingCap)) return deny(request, "CAP_EXCEEDED", now);
  const result: EvaluationResultV1 = { request, decision: "ALLOW", publicReasonClass: "OK", reservedAmount: request.amount,
    resultingCheckpoint: nextCheckpoint(request, request.amount, state.occurrenceCount + 1, now), resultNonce: request.requestId,
    attempt: request.attempt, issuedAt: now, expiry: request.expiry, machineId: ZERO_BYTES32, keyFingerprint: ZERO_BYTES32 };
  return result;
}

File: apps/fcc-extension/internal/protocol/evaluator.go
package protocol

import (
	"math/big"

	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/crypto"
)

const (
	DecisionDeny  uint8 = 0
	DecisionAllow uint8 = 1

	ReasonOK                    uint8 = 0
	ReasonPolicyDenied          uint8 = 1
	ReasonMalformed             uint8 = 2
	ReasonWrongDomain           uint8 = 3
	ReasonStaleInput            uint8 = 4
	ReasonDependencyUnavailable uint8 = 5
	ReasonExpired               uint8 = 6
	ReasonStopped               uint8 = 7
	ReasonInsufficientBalance   uint8 = 8
	ReasonCapExceeded           uint8 = 9
	ReasonOccurrenceExceeded    uint8 = 10
	ReasonTargetDenied          uint8 = 11
	ReasonRequesterDenied       uint8 = 12
	ReasonActionDenied          uint8 = 13
	ReasonFTSOInvalid           uint8 = 14
	ReasonCooldown              uint8 = 15

	ViolationPolicyDenied        uint64 = 1 << 0
	ViolationTargetDenied        uint64 = 1 << 1
	ViolationRequesterDenied     uint64 = 1 << 2
	ViolationActionDenied        uint64 = 1 << 3
	ViolationOccurrenceExceeded  uint64 = 1 << 4
	ViolationCooldown            uint64 = 1 << 5
	ViolationInsufficientBalance uint64 = 1 << 6
	ViolationFTSOInvalid         uint64 = 1 << 7
	ViolationCapExceeded         uint64 = 1 << 8
	policyViolationMaskV1               = ViolationPolicyDenied | ViolationTargetDenied | ViolationRequesterDenied | ViolationActionDenied | ViolationOccurrenceExceeded | ViolationCooldown | ViolationInsufficientBalance | ViolationFTSOInvalid | ViolationCapExceeded
)

// ComposePolicyDecisionV1 applies one fixed priority to simultaneous rule
// outcomes. The mask contains no policy plaintext or intermediate values.
func ComposePolicyDecisionV1(violations uint64) (uint8, uint8) {
	if violations & ^uint64(policyViolationMaskV1) != 0 {
		return DecisionDeny, ReasonMalformed
	}
	priorities := [...]struct {
		violation uint64
		reason    uint8
	}{
		{ViolationPolicyDenied, ReasonPolicyDenied},
		{ViolationTargetDenied, ReasonTargetDenied},
		{ViolationRequesterDenied, ReasonRequesterDenied},
		{ViolationActionDenied, ReasonActionDenied},
		{ViolationOccurrenceExceeded, ReasonOccurrenceExceeded},
		{ViolationCooldown, ReasonCooldown},
		{ViolationInsufficientBalance, ReasonInsufficientBalance},
		{ViolationFTSOInvalid, ReasonFTSOInvalid},
		{ViolationCapExceeded, ReasonCapExceeded},
	}
	for _, priority := range priorities {
		if violations&priority.violation != 0 {
			return DecisionDeny, priority.reason
		}
	}
	return DecisionAllow, ReasonOK
}

type FTSOSnapshotV1 struct {
	FeedID     common.Hash
	Value      *big.Int
	Decimals   uint8
	Timestamp  uint64
	Checkpoint common.Hash
}

type SpendStateV1 struct {
	AvailableBalance  *big.Int
	DailySpend        *big.Int
	RollingSpend      *big.Int
	OccurrenceCount   uint32
	LastExecutionAt   uint64
	SpendCheckpoint   common.Hash
	BalanceCheckpoint common.Hash
	Now               uint64
	FTSO              *FTSOSnapshotV1
}

func zeroHash() common.Hash { return common.Hash{} }

func denied(request ActionRequestV1, reason uint8, now uint64) EvaluationResultV1 {
	return EvaluationResultV1{Request: request, Decision: DecisionDeny, PublicReasonClass: reason, ReservedAmount: new(big.Int), ResultingCheckpoint: request.SpendCheckpoint, ResultNonce: request.RequestID, Attempt: request.Attempt, IssuedAt: now, Expiry: request.Expiry, MachineID: zeroHash(), KeyFingerprint: zeroHash()}
}

func containsAddress(values []common.Address, value common.Address) bool {
	for _, candidate := range values {
		if candidate == value {
			return true
		}
	}
	return false
}

func containsHash(values []common.Hash, value common.Hash) bool {
	for _, candidate := range values {
		if candidate == value {
			return true
		}
	}
	return false
}

// ReferenceValueV1 converts an asset amount to its reference value and rounds
// upward so a fractional unit cannot bypass a cap.
func ReferenceValueV1(amount, price *big.Int, decimals uint8) (*big.Int, bool) {
	if amount == nil || price == nil || amount.Sign() < 0 || amount.Cmp(maxUint256) > 0 || price.Sign() <= 0 || price.Cmp(maxUint256) > 0 || decimals > 36 {
		return nil, false
	}
	scale := new(big.Int).Exp(big.NewInt(10), new(big.Int).SetUint64(uint64(decimals)), nil)
	if amount.Sign() != 0 {
		limit := new(big.Int).Div(new(big.Int).Set(maxUint256), amount)
		if price.Cmp(limit) > 0 {
			return nil, false
		}
	}
	product := new(big.Int).Mul(amount, price)
	quotient, remainder := new(big.Int), new(big.Int)
	quotient.QuoRem(product, scale, remainder)
	if remainder.Sign() != 0 {
		quotient.Add(quotient, big.NewInt(1))
	}
	return quotient, true
}

func calculateNextCheckpoint(request ActionRequestV1, amount *big.Int, occurrence uint32, now uint64) (common.Hash, error) {
	requestHash, err := ActionRequestHash(request)
	if err != nil {
		return common.Hash{}, err
	}
	types, err := abi.NewType("bytes32", "", nil)
	if err != nil {
		return common.Hash{}, err
	}
	types2, err := abi.NewType("uint256", "", nil)
	if err != nil {
		return common.Hash{}, err
	}
	types3, err := abi.NewType("uint32", "", nil)
	if err != nil {
		return common.Hash{}, err
	}
	types4, err := abi.NewType("uint64", "", nil)
	if err != nil {
		return common.Hash{}, err
	}
	encoded, err := (abi.Arguments{{Type: types}, {Type: types}, {Type: types}, {Type: types2}, {Type: types3}, {Type: types4}}).Pack(SpendCheckpointTypeHash, request.SpendCheckpoint, requestHash, amount, occurrence, now)
	if err != nil {
		return common.Hash{}, err
	}
	return crypto.Keccak256Hash(encoded), nil
}

func EvaluatePolicy(policy PolicyV1, request ActionRequestV1, state SpendStateV1) (EvaluationResultV1, error) {
	normalized, err := normalizePolicy(policy)
	if err != nil {
		return EvaluationResultV1{}, err
	}
	commitment, err := PolicyCommitment(normalized)
	if err != nil {
		return EvaluationResultV1{}, err
	}
	if state.AvailableBalance == nil || state.DailySpend == nil || state.RollingSpend == nil || state.AvailableBalance.Sign() < 0 || state.DailySpend.Sign() < 0 || state.RollingSpend.Sign() < 0 || request.Amount == nil || request.Amount.Sign() < 0 {
		return denied(request, ReasonMalformed, state.Now), nil
	}
	if request.ChainID == nil || request.ChainID.Cmp(normalized.ChainID) != 0 || request.Registry != normalized.Registry || request.Vault != normalized.Vault || request.Router != normalized.Router || request.PolicyID != normalized.PolicyID || request.PolicyVersion != normalized.PolicyVersion || request.PolicyCommitment != commitment {
		return denied(request, ReasonWrongDomain, state.Now), nil
	}
	genesisCheckpoint, err := GenesisSpendCheckpoint(request.PolicyCommitment)
	if err != nil {
		return EvaluationResultV1{}, err
	}
	if request.SpendCheckpoint != state.SpendCheckpoint || request.BalanceCheckpoint != state.BalanceCheckpoint || state.OccurrenceCount == ^uint32(0) || request.Occurrence != state.OccurrenceCount+1 || (state.OccurrenceCount == 0 && request.SpendCheckpoint != genesisCheckpoint) {
		return denied(request, ReasonStaleInput, state.Now), nil
	}
	if request.Asset != normalized.Asset || request.ActionType != ActionFTestXRPTransfer || request.Amount.Sign() == 0 || request.CreatedAt > state.Now || request.Expiry < state.Now || request.GraceDeadline < request.CreatedAt || request.Expiry < request.GraceDeadline {
		if request.Expiry < state.Now {
			return denied(request, ReasonExpired, state.Now), nil
		}
		return denied(request, ReasonMalformed, state.Now), nil
	}
	var violations uint64
	if state.Now < normalized.StartAt || (normalized.EndAt != 0 && state.Now > normalized.EndAt) {
		violations |= ViolationPolicyDenied
	}
	if containsAddress(normalized.DenyTargets, request.Target) || (len(normalized.AllowTargets) > 0 && !containsAddress(normalized.AllowTargets, request.Target)) {
		violations |= ViolationTargetDenied
	}
	if len(normalized.AllowRequesters) > 0 && !containsAddress(normalized.AllowRequesters, request.Requester) {
		violations |= ViolationRequesterDenied
	}
	if len(normalized.AllowActionTypes) > 0 && !containsHash(normalized.AllowActionTypes, request.ActionType) {
		violations |= ViolationActionDenied
	}
	if normalized.MaxOccurrences != 0 && state.OccurrenceCount >= normalized.MaxOccurrences {
		violations |= ViolationOccurrenceExceeded
	}
	if normalized.CooldownSecs != 0 && state.LastExecutionAt != 0 && state.Now < state.LastExecutionAt+normalized.CooldownSecs {
		violations |= ViolationCooldown
	}
	if state.AvailableBalance.Cmp(request.Amount) < 0 {
		violations |= ViolationInsufficientBalance
	}
	referenceValue := new(big.Int).Set(request.Amount)
	if normalized.RequireFTSO {
		if state.FTSO == nil || state.FTSO.Value == nil || state.FTSO.FeedID != normalized.FTSOFeedID || state.FTSO.Timestamp > state.Now || state.Now-state.FTSO.Timestamp > normalized.MaxPriceAgeSecs || state.FTSO.Checkpoint == zeroHash() || request.InputCommitment != state.FTSO.Checkpoint {
			violations |= ViolationFTSOInvalid
		} else {
			value, valid := ReferenceValueV1(request.Amount, state.FTSO.Value, state.FTSO.Decimals)
			if !valid {
				violations |= ViolationFTSOInvalid
			} else {
				referenceValue = value
			}
		}
	}
	if violations&ViolationFTSOInvalid == 0 && ((normalized.MaxPerAction.Sign() != 0 && referenceValue.Cmp(normalized.MaxPerAction) > 0) || (normalized.DailyCap.Sign() != 0 && new(big.Int).Add(new(big.Int).Set(state.DailySpend), referenceValue).Cmp(normalized.DailyCap) > 0) || (normalized.RollingCap.Sign() != 0 && new(big.Int).Add(new(big.Int).Set(state.RollingSpend), referenceValue).Cmp(normalized.RollingCap) > 0)) {
		violations |= ViolationCapExceeded
	}
	decision, reason := ComposePolicyDecisionV1(violations)
	if decision == DecisionDeny {
		return denied(request, reason, state.Now), nil
	}
	nextCheckpoint, err := calculateNextCheckpoint(request, request.Amount, state.OccurrenceCount+1, state.Now)
	if err != nil {
		return EvaluationResultV1{}, err
	}
	return EvaluationResultV1{Request: request, Decision: DecisionAllow, PublicReasonClass: ReasonOK, ReservedAmount: new(big.Int).Set(request.Amount), ResultingCheckpoint: nextCheckpoint, ResultNonce: request.RequestID, Attempt: request.Attempt, IssuedAt: state.Now, Expiry: request.Expiry, MachineID: zeroHash(), KeyFingerprint: zeroHash()}, nil
}
